ext/EvoTreesCUDAExt/fit.jl

```diff
@@ -1,267 +1,330 @@
-
-    # initialize
-    n_current = [1]
-    depth = 1
-    jsg = CuVector(js)
+    epsv = eltype(tree_pred)(1e-8)
 
-    # initialize summary stats
-    copyto!(nodes[1].∑, sum(Float64, view(∇, :, nodes[1].is), dims=2))
+    @inbounds if depth < max_depth && best_gain[n_idx] > gamma
+        tree_split[node] = true
+        tree_cond_bin[node] = best_bin[n_idx]
+        tree_feat[node] = best_feat[n_idx]
+        tree_gain[node] = best_gain[n_idx]
 
-    # grow while there are remaining active nodes
-    while length(n_current) > 0 && depth <= params.max_depth
-        offset = 0 # identifies breakpoint for each node set within a depth
-        n_next = Int[]
+        child_l, child_r = node << 1, (node << 1) + 1
+        feat, bin = Int(tree_feat[node]), Int(tree_cond_bin[node])
 
-        if depth == params.max_depth
-            for n in n_current
-                if L <: Quantile
-                    pred_leaf_cpu!(tree.pred, n, nodes[n].∑, L, params, ∇, nodes[n].is)
-                else
-                    pred_leaf_cpu!(tree.pred, n, nodes[n].∑, L, params)
-                end
+        @inbounds for kk in 1:(2*K_val+1)
+            sum_val = zero(eltype(nodes_sum))
+            for b in 1:bin
+                sum_val += h∇[kk, b, feat, node]
             end
+            nodes_sum[kk, child_l] = sum_val
+            nodes_sum[kk, child_r] = nodes_sum[kk, node] - sum_val
+        end
+        
+        w_l = nodes_sum[2*K_val+1, child_l]
+        w_r = nodes_sum[2*K_val+1, child_r]
+        
+        if K_val == 1
+            g_l = nodes_sum[1, child_l]
+            h_l = nodes_sum[2, child_l]
+            nodes_gain[child_l] = g_l^2 / (h_l + lambda * w_l + epsv)
+            
+            g_r = nodes_sum[1, child_r]
+            h_r = nodes_sum[2, child_r]
+            nodes_gain[child_r] = g_r^2 / (h_r + lambda * w_r + epsv)
+            
+            tree_pred[1, child_l] = -g_l / (h_l + lambda * w_l + epsv)
+            tree_pred[1, child_r] = -g_r / (h_r + lambda * w_r + epsv)
         else
-            for n ∈ n_current[1:2:end]
-                update_hist!(nodes[n].h, h∇, ∇, x_bin, nodes[n].is, jsg)
-            end
-            @threads for n ∈ n_current[2:2:end]
-                if n % 2 == 0
-                    @views nodes[n].h[:, :, js] .= nodes[n>>1].h[:, :, js] .- nodes[n+1].h[:, :, js]
-                else
-                    @views nodes[n].h[:, :, js] .= nodes[n>>1].h[:, :, js] .- nodes[n-1].h[:, :, js]
-                end
-            end
-            sort!(n_current)
-            @threads for n ∈ n_current
-                update_gains!(L, nodes[n], js, params, feattypes, monotone_constraints)
+            gain_l = zero(eltype(nodes_gain))
+            gain_r = zero(eltype(nodes_gain))
+            
+            @inbounds for k in 1:K_val
+                g_l = nodes_sum[k, child_l]
+                h_l = nodes_sum[K_val+k, child_l]
+                gain_l += g_l^2 / (h_l + lambda * w_l / K_val + epsv)
```

Comment:
Why the divisions by `K_val`? There's no such operation in the current CPU or GPU implementations. Is there redundant addition of the weight that gets performed earlier?